lib/spack/spack/test/concretize_requirements.py

```diff
@@ -297,3 +297,43 @@ def test_requirements_are_higher_priority_than_deprecation(concretize_scope, tes
     s1 = Spec("y").concretized()
     assert s1.satisfies("@2.3")
     assert s1.satisfies("%gcc")
+
+
+@pytest.mark.parametrize("spec_str,requirement_str", [("x", "%gcc"), ("x", "%clang")])
+def test_default_requirements_with_all(spec_str, requirement_str, concretize_scope, test_repo):
+    """Test that users can override a deprecated version with a requirement."""
```

Comment:
The docstring for this and the next test need to be updated, in this case:

`Test that default requirements are applied to all packages`

and for the next one

`Test that specific package requirements override default package requirements`

also, it might be worth enforcing with `test_default_and_package_specific_requirements` that all default requirements are ignored for example if you have

```
require:
  all:
    require: %gcc ~shared
  y:
    require: "2.4"
```

then `y` will concretize to `+shared` (since that is the default for `y`).

Reply:
Should be done in 199f491